pkg/tenantctl/controller.go

```diff
@@ -207,18 +207,35 @@ func (c *Controller) syncHandler(key string) error {
 		if err != nil {
 			return err
 		}
-		tenant, err = c.removeFinalizerAndUpdate(tenant)
-		if err == nil {
-			c.syncCount++
+
+		isChanged, err := c.removeFinalizer(tenant)
+		if err != nil {
+			return err
+		}
+
+		if isChanged {
+			_, err = c.update(tenant)
+			if err == nil {
+				c.syncCount++
+			}
 		}
+
 		return err
 	}
 
-	tenant, err = c.addFinalizerAndUpdate(tenant)
+	isChanged, err := c.addFinalizer(tenant)
 	if err != nil {
 		return err
 	}
 
+	if isChanged {
+		_, err = c.update(tenant)
+		if err == nil {
+			c.syncCount++
+		}
+		return err
+	}
```

Comment:
Stopping the first reconciliation right after adding the finalizer entry would add a possibly significant delay until the tenant namespace gets available:

* Update events from the K8s API server may not be delivered instantly (depending on the API server's conditions), so that a new workqueue entry appears only lately.
* Many other tenant objects may be reconciled in between, e.g. due to a resync, while the current one is queued for a longer time.

Therefore the reconciliation should continue here.

The finalizer list _and_ the status section cannot be updated atomically. The finalizer list must be updated before the status to ensure the object cannot be deleted without clean-up. It's better to update the finalizer list _before_ calling `c.reconcile()`, so that in case this update fails we don't have to roll back (which may also fail and leaves garbage behind). This is how it was implemented before.

To avoid another unnecessary reconciliation due to a change _only in the finalizer list_, `(Controller).onTenantUpdate()` could be enhanced to sort out _unimportant_ changes. But it should still allow resync events (no changes at all, or only changes in unimportant fields). In case of no changes in important fields, it may drop the event only if the last status change happened recently, e.g. within half of the resync period.

Reply:
> In case of no changes in important fields, it may drop the event only if the last status change happened recently, e.g. within half of the resync period.

This doesn't work in case the event arrives while a (first) reconciliation is still in progress and has not written the status yet.

Reply:
Yes, for twice reconciliation, performance would be affected when many tenant objects created. However, update both finalizers and status section in once reconciliation not atomically makes the two tenant objects pointing to the same key name reconciled concurrently in some situation. (I explained above, but probably not it is not clear enough :) ). 

I think it would be better to fix this issue if finalizers and status section could be updated atomically or filter some `update` events in the method `(Controller).onTenantUpdate()`.